luarules/gadgets/unit_water_depth_damage.lua

```diff
@@ -0,0 +1,179 @@
+
+	if unitDef.health then
+		unitDefData[unitDefID].drowningDamage = unitDef.health * drowningDamage
+		unitDefData[unitDefID].fallDamage = unitDef.health * fallDamage * unitDefData[unitDefID].fallDamageMultiplier
+		unitDefData[unitDefID].unitDefID = unitDefID
+	end
+	if unitDef.moveDef.depth then
+		if unitDef.moveDef.depth >= isDrownableMaxWaterDepth then
+			if string.find(unitDef.moveDef.name, "hover") and not string.find(unitDef.moveDef.name, "raptor") then
+				unitDefData[unitDefID].isHover = true
+			else
+				unitDefData[unitDefID].isAmphibious = true
+			end
+		else 
+			unitDefData[unitDefID].isDrownable = true
+		end
+	end
+
+end
+
+function gadget:UnitLoaded(unitID, unitDefID, unitTeam,  transportID, transportTeam)
+	livingTransports[transportID] = true
+end
+
+function gadget:UnitUnloaded(unitID, unitDefID, unitTeam,  transportID, transportTeam)
+		fallingUnits[unitID] = fallingUnits[unitID] or {}
+		fallingUnits[unitID].transportID = transportID
+end
+
+function gadget:UnitLeftAir(unitID, unitDefID, unitTeam)
+	if fallingUnits[unitID] then
+		expiringFallingUnits[unitID] = gameFrame + gameFrameExpirationThreshold
+	end
+end
+
+function gadget:UnitEnteredWater(unitID, unitDefID, unitTeam)
+	if fallingUnits[unitID] and not livingTransports[fallingUnits[unitID].transportID] then
+		local velX, velY, velZ, velLength = spGetUnitVelocity(unitID)
+		local posX, posY, posZ = spGetUnitBasePosition(unitID)
+		if velLength > velocityThreshold then
+			spSpawnCEG('watersplash_large', posX, posY, posZ)
+			spPlaySoundFile('xplodep3', 0.5, posX, posY, posZ, 'sfx')
+			if unitDefData[unitDefID] then
+				local health, maxHealth = spGetUnitHealth(unitID)
+				local damage = (unitDefData[unitDefID].fallDamage * velLength) * (fallDamageCompoundingFactor ^ velLength)
+				if damage >= health then
+					spDestroyUnit(unitID) --this ensures a wreck is left behind. If damage is too great, it destroys the heap.
+				else
+				spAddUnitDamage(unitID, damage, 0, gaiaTeamID, waterDamageDefID)
+				end
+			end
+		else
+			spSpawnCEG('watersplash_small', posX, posY, posZ)
+			spPlaySoundFile('xplodep3', 0.3, posX, posY, posZ, 'sfx')
+		end
+		fallingUnits[unitID] = nil
+	else
+		fallingUnits[unitID] = nil
+	end
+
+	if unitDefData[unitDefID] and unitDefData[unitDefID].isDrownable then
+		drowningUnitsWatch[unitID] = unitDefData[unitDefID]
+	end
+end
+
+function gadget:UnitLeftWater(unitID, unitDefID, unitTeam)
+	drowningUnitsWatch[unitID] = nil
+end
+
+function gadget:UnitDestroyed(unitID, unitDefID, unitTeam, attackerID, attackerDefID, attackerTeam)
+	fallingUnits[unitID] = nil
+	expiringFallingUnits[unitID] = nil
+	livingTransports[unitID] = nil
+	drowningUnitsWatch[unitID] = nil
+end
+
+local function getUnitPositionHeight(unitID) -- returns nil for invalid units
+	if (spGetUnitIsDead(unitID) ~= false) or (spValidUnitID(unitID) ~= true) then return nil, nil, nil end
+	local posX, posY, posZ = spGetUnitPosition(unitID)
```

Comment:
what's the advantage of this function over just calling spGetUnitPosition?

Reply:
I stole this from behereth's old transport fall damage gadget. Without the checks for valid units, nil errors sometimes happen this filters them out